crates/sui/src/client_commands.rs

```diff
@@ -3098,3 +3107,66 @@ async fn check_protocol_version_and_warn(client: &SuiClient) -> Result<(), anyho
 
     Ok(())
 }
+
+/// Fetch move packages based on the provided package IDs.
+async fn fetch_move_packages(
+    client: &SuiClient,
+    package_ids: &[ObjectID],
+) -> Result<Vec<MovePackage>, anyhow::Error> {
+    let objects = client
+        .read_api()
+        .multi_get_object_with_options(package_ids.to_vec(), SuiObjectDataOptions::bcs_lossless())
+        .await
+        .map_err(|e| anyhow!("{e}"))?;
+
+    objects
+        .into_iter()
+        .map(|o| {
+            let o = o.into_object().map_err(|e| anyhow!("{e}"))?;
+            let Some(SuiRawData::Package(p)) = o.bcs else {
+                bail!("Expected SuiRawData::Package but got something else");
+            };
+            p.to_move_package(u64::MAX /* safe as this pkg comes from the network */)
+                .map_err(|e| anyhow!("{e}"))
+        })
+        .collect()
+}
+
+/// Filter out a package's dependencies which are not referenced in the source code.
+pub(crate) async fn pkg_tree_shake(
+    client: &SuiClient,
+    with_unpublished_dependencies: bool,
+    compiled_package: &mut CompiledPackage,
+) -> Result<(), anyhow::Error> {
+    let pkgs = compiled_package.find_immediate_deps_pkgs_to_keep(with_unpublished_dependencies)?;
+    let pkg_ids: Vec<_> = pkgs.clone().into_values().collect();
+
+    let pkg_name_to_orig_id: BTreeMap<_, _> = compiled_package
+        .package
+        .deps_compiled_units
+        .iter()
+        .map(|(pkg_name, module)| (pkg_name, ObjectID::from(module.unit.address.into_inner())))
+        .collect();
+
+    let published_deps_packages = fetch_move_packages(client, &pkg_ids).await?;
+    let linkage_table_ids = published_deps_packages
+        .iter()
+        .flat_map(|pkg| {
+            let linkage_table = pkg.linkage_table();
+            linkage_table.keys().cloned().collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
```

Comment:
- I think `flat_map` can deal with being passed an iterator? 
- Use a set -- you are checking for containment, and you don't want to deal with duplicates.

```suggestion
    let linkage_table_original_ids: BTreeSet<_> = published_deps_packages
        .iter()
        .flat_map(|pkg| pkg.linkage_table().keys())
        .cloned()
        .collect();
```